ext/lists.go

```diff
@@ -65,6 +89,26 @@ func (listsLib) CompileOptions() []cel.EnvOption {
 				}),
 			),
 		),
+		cel.Function("flatten",
+			cel.MemberOverload("list_flatten",
+				[]*cel.Type{listType}, listType,
```

Comment:
It seems somewhat unlikely that you'll get the right behavior at check-time for these signatures since the `list(T)` as input and `list(T)` as output could lead to situations where no reduction in type complexity occurs. You might need to specify `list(dyn)` to `list(dyn)` otherwise the subsequent expressions might choke at type-check time such as the example in #982

Reply:
@TristonianJones thanks for looking at it. I've changed the signatures to `list(dyn)` in https://github.com/google/cel-go/pull/980/commits/f12810c53371bc3d977817d650daabcdc96d149b

Reply:
@seirl Unfortunately this is a limit of the type-checker in that it can't reason about reduction in type-parameter complexity, especially when it's arbitrary, e.g. for `flatten(list(T), int) -> T`, how would the type-checker be able to infer the proper type of `T` when `T` could be `list(list(list(V)))` but the runtime output could be `list(V)` or some other permutation of a list.

The only other approach you could take would be to introduce something like the example output structure into the type-signature, but that's going to be tricky, e.g. `flatten(list(T), list(V)) -> list(V)`

Reply:
To @seirl's point if we added overloads for `flatten(list(list(T))) -> list(T)`, then we can get some stronger type-checking at least for depth of 1.